### PR TITLE
Rework search bar

### LINKED ISSUE #6625: Settings in Searching: suggest to separate these from autocompletion
JabRef 5.0 under Windows 10.

Usability of Jabref is reduced for me by the following issue related to searching (i.e. searching for  entries within the current library).

 The options for using RegEx or not and searching case-sensitive or not are only available via Auto-complete:
Untitled
. I don't see the logic behind that; having these available adjacent to the search bar, irrespective of using autocomplete or not, (as it used to be, see below) makes more sense to me. 

search-bar

### PR DESCRIPTION
**Pull request by calixtus**
picturemessage_3ata4qts chg

fixes #6625 

- [x] Change in CHANGELOG.md described (if applicable)
- [ ] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [x] Screenshots added in PR description (for UI changes)
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, submitted a pull request to the documentation repository.

### CODE REVIEW

File: src/main/java/org/jabref/gui/search/GlobalSearchBar.java [hunk 1]
```diff
@@ -325,7 +322,7 @@ private void updateResults(int matched, TextFlow description, boolean grammarBas
     }
 
     private void setHintTooltip(TextFlow description) {
-        if (Globals.prefs.getBoolean(JabRefPreferences.SHOW_ADVANCED_HINTS)) {
+        if (preferencesService.getGeneralPreferences().shouldShowAdvancedHints()) {
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/gui/search/GlobalSearchBar.java:325]: I think it would make sense to add it to the search preferences as well

Comment (calixtus) [reply src/main/java/org/jabref/gui/search/GlobalSearchBar.java:334]: Originally the idea about the advanced hints options was to make it reusable for other hints as well. But in fact, it has only been used also for another (somewhat silly) hint in the network tab in the preferences dialog. So im not sure about this... Maybe in the future, we could evolve that feature to some annoying tip of the day feature? :wink:

Comment (Siedlerchr) [reply src/main/java/org/jabref/gui/search/GlobalSearchBar.java:334]: I have no strong opinion on this. So then just leave it as is

File: src/main/java/org/jabref/preferences/SearchPreferences.java [hunk 2]
```diff
@@ -1,103 +1,63 @@
...
-        preferences.putBoolean(SEARCH_REG_EXP, isRegularExpression);
-        return this;
+    public Builder getBuilder() {
+        return new Builder(this);
     }
 
-    public int getSeachDialogWidth() {
-        return preferences.getInt(SEARCH_DIALOG_WIDTH);
-    }
+    public static class Builder {
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/preferences/SearchPreferences.java:33]: Is the builder really necessary? I think the `with*` methods can be added directly to the main class without any problems.

**Comment by calixtus** [comment]
> The changes look good to me. The only concern that I have is that the search bar is now really small. Thus, while editing a more complex query, it is no longer possible to see the complete query (and "scrolling" in a text field is rather tedious). This was the main reason for the "extend on focus" mode. I'm not particularly attached to the extend animation but it should be still possible to comfortably edit more complex queries.

I'm not quite sure I understand what you mean about the small search bar... I removed the previous restriction of the searchbar to expand to the right, so now it's always extended. The empty space still showing between the searchbar and the toolbar buttons is left over from before and shows "Found X entries". I did not yet had a good idea what to do with this message, so i did not touch it (except adding 4 pixels distance to the searchbar).

For comparison: This is how the master branch looks right now:
grafik

File: src/main/java/org/jabref/preferences/SearchPreferences.java (hunk 2 above)
Comment (calixtus) [reply src/main/java/org/jabref/preferences/SearchPreferences.java:33]: I'll change it, I originally just wanted to stick close to the builder pattern, as I don't have much experience on it, so this was somewhat a study project for me :wink:

**Comment by tobiasdiez** [comment]
Ah ok, then I misread the code and the first screenshot - I thought that's now the size of the search bar.

File: src/main/java/org/jabref/preferences/JabRefPreferences.java [hunk 3]
```diff
@@ -2356,4 +2363,27 @@ public void storeSpecialFieldsPreferences(SpecialFieldsPreferences specialFields
...
     }
+
+    @Override
+    public SearchPreferences getSearchPreferences() {
+        SearchDisplayMode searchDisplayMode;
+        try {
+            searchDisplayMode = SearchDisplayMode.valueOf(get(SEARCH_DISPLAY_MODE));
+        } catch (IllegalArgumentException ex) {
+            // Should only occur when the searchmode is set directly via preferences.put and the enum was not used
+            searchDisplayMode = SearchDisplayMode.valueOf((String) defaults.get(SEARCH_DISPLAY_MODE));
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/preferences/JabRefPreferences.java:2374]: The get() Method already returns the default if no value:
```
 public String get(String key) {
        return prefs.get(key, (String) defaults.get(key));
    }
```

Comment (calixtus) [reply src/main/java/org/jabref/preferences/JabRefPreferences.java:2374]: Good catch. I just copied the logic from the former searchpreferences.class. Fixing the missing consistency is one of the big projects in the future...

Comment (Siedlerchr) [reply src/main/java/org/jabref/preferences/JabRefPreferences.java:2374]: This was the only issue I found so far,the rest looks good!

Comment (calixtus) [reply src/main/java/org/jabref/preferences/JabRefPreferences.java:2374]: Hmm... I thought about this: what if the returned stored value is defect, but not the default value?

Comment (Siedlerchr) [reply src/main/java/org/jabref/preferences/JabRefPreferences.java:2374]: Why should that happen?  the Prefernces.get says:
>the value associated with key, or def if no value is associated with key, or the backingstore is inaccessible.

**Pull request merged**